contracts/utils/NameEncoder.sol

```diff
@@ -0,0 +1,104 @@
+pragma solidity ^0.8.4;
+
+import "./Strings.sol";
+
+library NameEncoder {
+    using strings for *;
+
+    function encodePart(strings.slice memory name)
+        internal
+        view
+        returns (bytes memory)
+    {
+        bytes memory encodedName;
+        strings.slice memory label = name.split(".".toSlice());
+        if (label.empty()) {
+            return abi.encodePacked(uint8(0));
+        }
+        encodedName = encodePart(name);
+
+        return
+            abi.encodePacked(
+                uint8(label.len()),
+                bytes(label.toString()),
+                encodedName
+            );
+    }
+
+    function encode(string memory name) internal view returns (bytes memory) {
+        strings.slice memory sliceName = name.toSlice();
+        if (sliceName.empty()) {
+            return abi.encodePacked(uint8(0), uint8(0));
+        }
+        return encodePart(sliceName);
+    }
+
+    function encodeAndHash(string memory name)
+        internal
+        view
+        returns (bytes memory, bytes32)
+    {
+        bytes memory encodedName = encode(name);
+        return (encodedName, namehash(encodedName, 0));
```

Comment:
It'd be more efficient to produce the hash at the same time as converting the name.